### ISSUE TITLE
Forces Calculation - Derivatives

### ISSUE THREAD

**Issue by CayetanaZF** [openmm/openmm#4488]
Hey guys,

Hoping everyone's fine! 
Maybe I'm asking a dumb question but I'm struggling with the explicit calculation of the energy derivatives and how OpenMM internally calculate it. For example, a harmonic bond force potential, 0.5*k*(x-xi)**2, where k is a constant and the potential function only depend on the distance between two atoms (x) and its reference (xi). It's magnitude would be k*(x-xi), but when I ask OpenMM to get the forces, it handed to me a matrix with the force component by spatial vector (xyz), so how it goes from a distance to a vector variable?  Also, how do I get the magnitude of the total force of that potencial (calculates the norm of every vectorial set per atom and then add it )?

Best regard, 

Caye!

**Comment by peastman** [comment]
Each component of the force on an atom is the negative derivative of the energy with respect to that degree of freedom.  For example, the x component of the force on atom 1 from a bond is given by $-\frac{dE}{dr}\frac{\partial r}{\partial x_1}$.  If the bond connects atoms 1 and 2, then $r = \sqrt{(x_1-x_2)^2 + (y_1-y_2)^2 + (z_1-z_2)^2}$.  Taking the derivative gives $\frac{\partial r}{\partial x_1} = \frac{x_1-x_2}{r}$.

>  Also, how do I get the magnitude of the total force of that potencial (calculates the norm of every vectorial set per atom and then add it )?

The magnitude of a vector $\mathbf{f}$ is $\sqrt{f_x^2+f_y^2+f_z^2}$.  Or is that not what you were asking?

**Comment by CayetanaZF** [comment]
> The magnitude of a vector f is fx2+fy2+fz2. Or is that not what you were asking?

Yes, I was asking about that thank you! But also in relation to the total force magnitude over the whole set of atoms. Would it be just the sum of every force magnitude per atom? And dear @peastman, if it weren't too much trouble, what would it be the dihedral xyz parametrization internally use to calculate dihedrals?

Best regard, 

Caye!

**Comment by peastman** [comment]
> Would it be just the sum of every force magnitude per atom?

Sum the squares of all force components and take the square root.

> what would it be the dihedral xyz parametrization internally use to calculate dihedrals?

What do you mean?

**Comment by CayetanaZF** [comment]
> Sum the squares of all force components and take the square root.

Sure, thank you!

> What do you mean?

As the distance between two points, `r = sqrt((x_{1}-x_{2})^{2}+(y_{1}-y_{2})^{2}+(z_{1}-z_{2})^{2})`, what's the cartesian representation of a dihedral that openMM uses?

**Comment by peastman** [comment]
There's a fair amount of complexity to calculating a dihedral, especially because there are singularities you need to be careful of.  [This file](https://github.com/openmm/openmm/blob/master/platforms/common/src/kernels/torsionForce.cc) contains code for computing forces based on a dihedral.
